rfcs/2021-04-13-7027-core-extraction.md

```diff
@@ -0,0 +1,319 @@
+
+This alternative is the inverse of the previous: core is just another
+package. We extract what code is "core" to vector and move it into its own
+package, rigged with private tests, interfaces in the manner of VRL. This will
+allow work on core to happen in a similar manner to VRL today, with tighter
+iteration and more focus. Experimentation in core specifically need not include
+the rest of the project, until after the experiment is shown to be
+positive. Casual contributors will be unlikely to work with core -- most outside
+contributions have historically been sources or sinks -- but, if they do, we'll
+have a less resource intensive package for them to work against. Core as just
+another package lends itself to an iterative approach: extraction of core pieces
+need not happen in one burst but bit by bit. This approach has the added benefit
+of deferring questions of non-core project structure, how our feature flags are
+set up etc. There will be some grey area of course -- I argue that configuration
+is not a core concern but am sympathetic to the counter -- but these can be
+resolved in the piecemeal process of moving things into "core".
+
+## Proposal
+
+I argue that "Core is Just Another Package" is the best path forward. This
+approach addresses each of our major goals for extracting core without imposing
+significant secondary questions about total project organization, of which there
+has been nervous concern in the team. We don't want to "cut the project
+boundaries wrong" and I share this concern. I think we will eventually want more
+rigorous structure in the non-core parts of the project but I would like to see
+that deferred.
+
+## Plan of Action
+
+To extract the core of vector I propose that we do the following:
+
+* Create a new, blank `core` package that top*level vector depends on.
+* Migrate one of the areas from 'What is "core" to vector?' into this package,
+  along with any existing test code.
+* Add any new test code that seems worthwhile for the migrated material.
+* Repeat.
+* Once a critical mass of core exists, create a core-private source for
+  generating `Event` load into a topology and a core-private sink that signals
+  for program termination after a set number of `Event`s have been
+  received. Bundle this into a `vector*core` for throughput trials of core only,
+  validating the package structure for cheap experimentation and for avoiding
+  core as "vector but with pieces missing".
+* Migrate any remaining core concepts.
+
+By the end of this process we will have a "core" package that represents all the
+pieces needed to build the backbone of a vector, used in the top-level vector
+package but independently able to be rigged into test harnesses and experimented
+with in a manner that does not necessarily require chasing changes through the
+project tree. We can then start the process of pushing core to have perfect
+mechanical sympathy, in line with our goals in RFC 6531.
```

Comment:
I can imagine this working very well for improving the absolute core components like event representation and topology communication, but I'm curious what our next steps would be.

When we want to expand the scope of this work to things like VRL, will we go back to working in top-level Vector or do you imagine a path of gradually expanding what we consider "core"? I can see us wanting to do the same level of deep iterative improvement to a larger and larger subset of components over time.

It's totally fine if that's out of scope for this first piece of work, I'm just curious how you're thinking about it.

Reply:
Real talk: I don't know. My hope is that once we've got "core" pretty well separated as an individual thing we'll see broader organization principles we can play with, stuff I have _notions_ about today but which are pretty pie in the sky. I kind of believe there'll be a Rust `core`/`alloc`/`std` analog that happens with vector where the initial distinction is not clear, then `core`/`std` is realized as A Thing and subsequently `alloc` is also pulled out because there's a use-case for just `core`+`alloc`. Strikes me it'll take time and discussion to realize something similar in vector, but that's my sense of where things will go.